provider/pihole/clientV6.go

```diff
@@ -0,0 +1,376 @@
+}
+
+func (p *piholeClientV6) deleteRecord(ctx context.Context, ep *endpoint.Endpoint) error {
+	return p.apply(ctx, http.MethodDelete, ep)
+}
+
+func (p *piholeClientV6) aRecordsScript() string {
+	return fmt.Sprintf("%s/api/config/dns/hosts", p.cfg.Server)
+}
+
+func (p *piholeClientV6) cnameRecordsScript() string {
+	return fmt.Sprintf("%s/api/config/dns/cnameRecords", p.cfg.Server)
+}
+
+func (p *piholeClientV6) urlForRecordType(rtype string) (string, error) {
+	switch rtype {
+	case endpoint.RecordTypeA, endpoint.RecordTypeAAAA:
+		return p.aRecordsScript(), nil
+	case endpoint.RecordTypeCNAME:
+		return p.cnameRecordsScript(), nil
+	default:
+		return "", fmt.Errorf("unsupported record type: %s", rtype)
+	}
+}
+
+// ApiAuthResponse Define a struct to match the JSON response /auth/app structure
+type ApiAuthResponse struct {
+	Session struct {
+		Valid    bool   `json:"valid"`
+		TOTP     bool   `json:"totp"`
+		SID      string `json:"sid"`
+		CSRF     string `json:"csrf"`
+		Validity int    `json:"validity"`
+		Message  string `json:"message"`
+	} `json:"session"`
+	Took float64 `json:"took"`
+}
+
+// ApiErrorResponse Define struct to match the JSON structure
+type ApiErrorResponse struct {
+	Error struct {
+		Key     string `json:"key"`
+		Message string `json:"message"`
+		Hint    string ` json:"hint"`
+	} `json:"error"`
+	Took float64 `json:"took"`
+}
+
+// ApiRecordsResponse Define struct to match JSON structure
+type ApiRecordsResponse struct {
+	Config struct {
+		DNS struct {
+			Hosts        []string `json:"hosts"`
+			CnameRecords []string `json:"cnameRecords"`
+		} `json:"dns"`
+	} `json:"config"`
+	Took float64 `json:"took"`
+}
+
+func (p *piholeClientV6) apply(ctx context.Context, action string, ep *endpoint.Endpoint) error {
+	if !p.cfg.DomainFilter.Match(ep.DNSName) {
+		log.Debugf("Skipping %s %s that does not match domain filter", action, ep.DNSName)
+		return nil
+	}
+	apiUrl, err := p.urlForRecordType(ep.RecordType)
+	if err != nil {
+		log.Warnf("Skipping unsupported endpoint %s %s %v", ep.DNSName, ep.RecordType, ep.Targets)
+		return nil
+	}
+
+	if p.cfg.DryRun {
+		log.Infof("DRY RUN: %s %s IN %s -> %s", action, ep.DNSName, ep.RecordType, ep.Targets[0])
+		return nil
+	}
+
+	log.Infof("%s %s IN %s -> %s", action, ep.DNSName, ep.RecordType, ep.Targets[0])
+
+	// Get the current record
+	if strings.Contains(ep.DNSName, "*") {
+		return provider.NewSoftError(errors.New("UNSUPPORTED: Pihole DNS names cannot return wildcard"))
+	}
+
+	switch ep.RecordType {
+	case endpoint.RecordTypeA, endpoint.RecordTypeAAAA:
+		apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s %s", ep.Targets, ep.DNSName)))
+	case endpoint.RecordTypeCNAME:
+		if ep.RecordTTL.IsConfigured() {
+			apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s,%s,%d", ep.DNSName, ep.Targets, ep.RecordTTL)))
+		} else {
+			apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s,%s", ep.DNSName, ep.Targets)))
+		}
+	}
```

Comment:
we could have helper fucntion to generate apiUrl function, to avoid repeaing same stuff

```go
func (p *piholeClientV6) generateApiUrl(baseUrl, params string) string {
	return fmt.Sprintf("%s/%s", baseUrl, url.PathEscape(params))
}
```